### PR TITLE
PHP 7 support

### PR DESCRIPTION
**Pull request by alexeyshockov**
This is initial support for PHP 7 (that's available in 2.0 version of great PHP parser by nikic).

I haven't found any plans for the topic, so I've decided to create a PR.

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
:-1:

**Comment by GrahamCampbell** [comment]
We should retain compatibility with both 1.x and 2.x.

File: composer.json [hunk 1]
```diff
@@ -13,10 +13,10 @@
         }
     ],
     "require": {
-        "php": ">=5.3.9",
+        "php": ">=5.4",
```
Comment (GrahamCampbell) [inline composer.json]: Should be reverted.

File: src/Psy/CodeCleaner/ValidFunctionNamePass.php [hunk 2]
```diff
@@ -58,7 +59,7 @@ public function leaveNode(Node $node)
         if ($node instanceof FuncCall) {
             // if function name is an expression or a variable, give it a pass for now.
             $name = $node->name;
-            if (!$name instanceof Expression && !$name instanceof Variable) {
+            if (!$name instanceof Expr && !$name instanceof Variable) {
```
Comment (GrahamCampbell) [inline src/Psy/CodeCleaner/ValidFunctionNamePass.php]: ?

File: src/Psy/Command/ParseCommand.php [hunk 3]
```diff
@@ -126,7 +127,9 @@ private function parse($code)
     private function getParser()
     {
         if (!isset($this->parser)) {
-            $this->parser = new Parser(new Lexer());
+            $this->parser = (new ParserFactory())->create(
+                version_compare(PHP_VERSION, '7.0', '>=') ? ParserFactory::ONLY_PHP7 : ParserFactory::ONLY_PHP5
```
Comment (GrahamCampbell) [inline src/Psy/Command/ParseCommand.php]: ?

Comment (GrahamCampbell) [inline src/Psy/Command/ParseCommand.php]: just use the prefer php 7 one, and be done with it surely?

File: src/Psy/Command/ParseCommand.php [hunk 4]
```diff
@@ -14,6 +14,7 @@
 use PhpParser\Lexer;
```
Comment (GrahamCampbell) [inline src/Psy/Command/ParseCommand.php]: this is now an unused import

**Comment by GrahamCampbell** [comment]
Something like this would be fine: https://github.com/jeremeamia/super_closure/pull/70/files.

File: src/Psy/CodeCleaner/ValidFunctionNamePass.php (hunk 2 above)
Comment (alexeyshockov) [inline src/Psy/CodeCleaner/ValidFunctionNamePass.php]: I think it's a bug, because PHP parser doesn't have `Expression` type. It has only `Expr`.

**Comment by alexeyshockov** [comment]
I agree, that support for both versions of PHP parser is better.

File: src/Psy/Command/ParseCommand.php (hunk 3 above)
Comment (alexeyshockov) [inline src/Psy/Command/ParseCommand.php]: In the `CodeCleaner` strict options should be used, because parsed code will be evaluated on the same interpreter (so with `PREFER_PHP7` invalid code can be passed to 5.x interpreter).

I don't have strong opinion about `ParseCommand`. I think the best way will be to add additional console option for PHP version. What do you think?

File: src/Psy/CodeCleaner/ValidFunctionNamePass.php (hunk 2 above)
Comment (GrahamCampbell) [inline src/Psy/CodeCleaner/ValidFunctionNamePass.php]: That should be fixed in a separate PR then.

File: src/Psy/Command/ParseCommand.php [hunk 5]
```diff
@@ -126,7 +126,9 @@ private function parse($code)
     private function getParser()
     {
         if (!isset($this->parser)) {
-            $this->parser = new Parser(new Lexer());
+            $this->parser = (new ParserFactory())->create(
```
Comment (GrahamCampbell) [inline src/Psy/Command/ParseCommand.php]: this class doesn't exist on 1.x

Comment (GrahamCampbell) [inline src/Psy/Command/ParseCommand.php]: you need to update this

Comment (GrahamCampbell) [inline src/Psy/Command/ParseCommand.php]: you can't just update the version constraint to allow 1.x again, but leave the code here that only works on 2.x

Comment (alexeyshockov) [inline src/Psy/Command/ParseCommand.php]: I already understood, don't work as CI ;)

File: src/Psy/CodeCleaner/ValidFunctionNamePass.php (hunk 2 above)
Comment (alexeyshockov) [inline src/Psy/CodeCleaner/ValidFunctionNamePass.php]: Agree. Moved to a separated PR (#252). Thanks!

File: src/Psy/Command/ParseCommand.php (hunk 4 above)
Comment (alexeyshockov) [inline src/Psy/Command/ParseCommand.php]: Now it's used again (after reverting support for PHP parser 1.x).

**Comment by alexeyshockov** [comment]
So, tests pass. Only one question about proper option to PHP parser (`PREFER_*` or `ONLY_*`) is opened now, as I can see.

File: src/Psy/CodeCleaner.php [hunk 6]
```diff
@@ -53,7 +54,17 @@ class CodeCleaner
      */
     public function __construct(Parser $parser = null, Printer $printer = null, NodeTraverser $traverser = null)
     {
-        $this->parser    = $parser    ?: new Parser(new Lexer());
+        if (class_exists('PhpParser\ParserFactory')) {
```
Comment (GrahamCampbell) [inline src/Psy/CodeCleaner.php]: this code is duplicated - we need to find a way to avoid that

**Comment by GrahamCampbell** [comment]
Maybe the other class alias fixes should be moved to another PR too?

**Comment by GrahamCampbell** [comment]
Maybe just make them two commits and keep them both in here, then just rebase if the other PRs are merged?

**Comment by alexeyshockov** [comment]
Splited to two commits.

**Comment by bobthecow** [comment]
Thanks for getting this going!

Let's make our own parser factory of some sort so we don't duplicate the parser creation logic. It should probably have an optional "support php 7" flag, which defaults to "is this currently php 7?" but could be passed as an arg from the parse command.

**Comment by alexeyshockov** [comment]
Good, I added own factory. It just forwards original PHP parser factory constants, because `parse` command doesn't abstract user over the library, it's only an interface.

File: src/Psy/CodeCleaner.php [hunk 7]
```diff
@@ -53,7 +52,9 @@ class CodeCleaner
      */
     public function __construct(Parser $parser = null, Printer $printer = null, NodeTraverser $traverser = null)
     {
-        $this->parser    = $parser    ?: new Parser(new Lexer());
+        $parserFactory = new ParserFactory();
```
Comment (GrahamCampbell) [inline src/Psy/CodeCleaner.php]: hmmm, only create the parser factory if we need it surely?

Comment (GrahamCampbell) [inline src/Psy/CodeCleaner.php]: ie, `$this->parser    = $parser    ?: (new ParserFactory())->createParser();`

Comment (alexeyshockov) [inline src/Psy/CodeCleaner.php]: Sad, but not working with PHP 5.3 :(

Comment (GrahamCampbell) [inline src/Psy/CodeCleaner.php]: we should add an if statement then

Comment (bobthecow) [inline src/Psy/CodeCleaner.php]: But this works on 5.3 :)

```php
        if ($parser === null) {
            $parserFactory = new ParserFactory();
            $parser = $parserFactory->createParser();
        }

        $this->parser    = $parser;
```

Comment (GrahamCampbell) [inline src/Psy/CodeCleaner.php]: > But this works on 5.3 :)

Yeh. :P

File: src/Psy/Command/ParseCommand.php [hunk 8]
```diff
@@ -59,12 +68,23 @@ public function setPresenter(Presenter $presenter)
...
     {
+        $definition = array(
+            new InputArgument('code', InputArgument::REQUIRED, 'PHP code to parse.'),
+            new InputOption('depth', '', InputOption::VALUE_REQUIRED, 'Depth to parse', 10),
+        );
+        if ($this->parserFactory->hasKindsSupport()) {
+            $definition[] = new InputOption(
+                'parser-kind',
+                '',
+                InputOption::VALUE_OPTIONAL,
```
Comment (bobthecow) [inline src/Psy/Command/ParseCommand.php]: This should be `VALUE_REQUIRED`.

File: src/Psy/Command/ParseCommand.php [hunk 9]
```diff
@@ -59,12 +68,23 @@ public function setPresenter(Presenter $presenter)
      */
     protected function configure()
     {
+        $definition = array(
+            new InputArgument('code', InputArgument::REQUIRED, 'PHP code to parse.'),
+            new InputOption('depth', '', InputOption::VALUE_REQUIRED, 'Depth to parse', 10),
+        );
+        if ($this->parserFactory->hasKindsSupport()) {
+            $definition[] = new InputOption(
+                'parser-kind',
```
Comment (bobthecow) [inline src/Psy/Command/ParseCommand.php]: I'm not sure I love `parser-kind` as an option name. I don't have a better suggestion right now, other than to drop the `parser-` prefix, since that's kind of implied by the command :)

File: src/Psy/ParserFactory.php [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    /**
+     * @return string|null
+     */
+    public function getDefaultKind()
+    {
+        if (self::hasKindsSupport()) {
+            return version_compare(PHP_VERSION, '7.0', '>=') ? self::ONLY_PHP7 : self::ONLY_PHP5;
+        } else {
```
Comment (bobthecow) [inline src/Psy/ParserFactory.php]: the `else return` is unnecessary.

File: src/Psy/ParserFactory.php [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+            $kind = $kind ?: self::getDefaultKind();
+
+            if (!in_array($kind, self::getPossibleKinds())) {
+                throw new \InvalidArgumentException('Unknown parser kind');
+            }
+
+            $parser = $originalFactory->create(constant('PhpParser\ParserFactory::' . $kind));
+        } else {
+            if ($kind !== null) {
+                throw new \InvalidArgumentException('You must install second version of PHP parser to specify kind.');
```
Comment (bobthecow) [inline src/Psy/ParserFactory.php]: `Install PHP Parser v2.x to specify parser kind` or something like that?

File: src/Psy/ParserFactory.php [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+    /**
+     * Available kinds for second version of PHP parser.
+     *
+     * @return array
+     */
+    private function getPossibleKinds()
+    {
+        $factory = new ReflectionClass(__CLASS__);
+
+        return array_values($factory->getConstants());
```
Comment (bobthecow) [inline src/Psy/ParserFactory.php]: Why don't we add a private static property rather than using reflection for something that will almost never change?

File: src/Psy/Command/ParseCommand.php [hunk 13]
```diff
@@ -59,12 +68,23 @@ public function setPresenter(Presenter $presenter)
      */
     protected function configure()
     {
+        $definition = array(
+            new InputArgument('code', InputArgument::REQUIRED, 'PHP code to parse.'),
+            new InputOption('depth', '', InputOption::VALUE_REQUIRED, 'Depth to parse', 10),
+        );
```
Comment (bobthecow) [inline src/Psy/Command/ParseCommand.php]: Add a newline here for readability.

File: src/Psy/Command/ParseCommand.php [hunk 14]
```diff
@@ -59,12 +68,23 @@ public function setPresenter(Presenter $presenter)
...
+        if ($this->parserFactory->hasKindsSupport()) {
+            $definition[] = new InputOption(
+                'parser-kind',
+                '',
+                InputOption::VALUE_OPTIONAL,
+                "One PhpParser\\ParserFactory constants (default is based on current interpreter's version)",
+                $this->parserFactory->getDefaultKind()
```
Comment (bobthecow) [inline src/Psy/Command/ParseCommand.php]: I'd like to enumerate the possible parser kinds, probably in the `<<<HELP` block below? Also wouldn't hurt to add an example while we're at it :)

File: src/Psy/ParserFactory.php (hunk 11 above)
Comment (alexeyshockov) [inline src/Psy/ParserFactory.php]: Agree. Fixed.

File: src/Psy/ParserFactory.php (hunk 10 above)
Comment (alexeyshockov) [inline src/Psy/ParserFactory.php]: Fixed.

File: src/Psy/Command/ParseCommand.php (hunk 14 above)
Comment (alexeyshockov) [inline src/Psy/Command/ParseCommand.php]: I added all possibilities to the option description. I think, that's enough :)

File: src/Psy/Command/ParseCommand.php (hunk 8 above)
Comment (alexeyshockov) [inline src/Psy/Command/ParseCommand.php]: Fixed.

File: src/Psy/Command/ParseCommand.php (hunk 9 above)
Comment (alexeyshockov) [inline src/Psy/Command/ParseCommand.php]: Agree with simple `kind`. Fixed.

File: src/Psy/Command/ParseCommand.php (hunk 13 above)
Comment (alexeyshockov) [inline src/Psy/Command/ParseCommand.php]: Fixed.

File: src/Psy/CodeCleaner.php (hunk 7 above)
Comment (alexeyshockov) [inline src/Psy/CodeCleaner.php]: Agree, fixed to `if`. Hope we will see PHP version increment in the requirements soon...

File: src/Psy/ParserFactory.php (hunk 12 above)
Comment (alexeyshockov) [inline src/Psy/ParserFactory.php]: I don't like to duplicate things this way (yet another string with the same constant names). But it's not so important for me, so if you think differently...

File: src/Psy/ParserFactory.php [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+     * @return array
+     */
+    public static function getPossibleKinds()
+    {
+        $factory = new ReflectionClass(__CLASS__);
+
+        return array_values($factory->getConstants());
+    }
+
+    /**
```
Comment (GrahamCampbell) [inline src/Psy/ParserFactory.php]: missing short description

File: src/Psy/ParserFactory.php [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+     * @return bool
+     */
+    public function hasKindsSupport()
+    {
+        return class_exists('PhpParser\ParserFactory');
+    }
+
+    /**
```
Comment (GrahamCampbell) [inline src/Psy/ParserFactory.php]: missing short description

File: src/Psy/ParserFactory.php [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+     * @return string|null
+     */
+    public function getDefaultKind()
+    {
+        if ($this->hasKindsSupport()) {
+            return version_compare(PHP_VERSION, '7.0', '>=') ? static::ONLY_PHP7 : static::ONLY_PHP5;
+        }
+    }
+
+    /**
```
Comment (GrahamCampbell) [inline src/Psy/ParserFactory.php]: missing short description

File: src/Psy/Command/ParseCommand.php [hunk 18]
```diff
@@ -28,7 +28,16 @@
 class ParseCommand extends Command implements PresenterAware
 {
     private $presenter;
-    private $parser;
+    private $parserFactory;
+    private $parsers;
+
+    public function __construct($name = null)
```
Comment (GrahamCampbell) [inline src/Psy/Command/ParseCommand.php]: some phpdoc would be nice here so we know what name actually is

**Comment by GrahamCampbell** [comment]
Asside from a few more minor things I've just commented, this is looking pretty good now. :+1:

File: src/Psy/CodeCleaner.php (hunk 7 above)
Comment (bobthecow) [inline src/Psy/CodeCleaner.php]: > Hope we will see PHP version increment in the requirements soon...

Probably not, at least until it becomes too arduous to support. PsySH is useful even for people whose legacy support needs preclude them from using PHP 5.4+ :)

File: src/Psy/ParserFactory.php (hunk 12 above)
Comment (bobthecow) [inline src/Psy/ParserFactory.php]: I'd prefer changing it. I don't like reflection at runtime any more than we have to.

And yeah, I know, I'm the creator of a REPL that does an obscene amount of runtime reflection :)

File: src/Psy/Command/ParseCommand.php (hunk 18 above)
Comment (bobthecow) [inline src/Psy/Command/ParseCommand.php]: just use an inheritDoc tag :)

**Comment by bobthecow** [comment]
:+1:

Fill in the missing docs that @GrahamCampbell pointed out, and change that one bit of reflection, and we're good to merge. Thanks!

**Comment by bobthecow** [comment]
@alexeyshockov around? i've got a bit of time right now, and won't have a chance to look at this again for a week or so, so i wanted to merge soon :)

**Comment by alexeyshockov** [comment]
I'm going to send updates later today, at night. I hope it will not be late!

On Saturday, November 7, 2015, Justin Hileman <notifications@github.com>
wrote:

> @alexeyshockov <https://github.com/alexeyshockov> around? i've got a bit
> of time right now, and won't have a chance to look at this again for a week
> or so, so i wanted to merge soon :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bobthecow/psysh/pull/251#issuecomment-154718813>.
>


-- 
Alexey Shockov

tel:+79652121992
skype:alexeyshockov
mailto:alexey@shockov.com
http://alexey.shockov.com/

File: src/Psy/ParserFactory.php (hunk 16 above)
Comment (alexeyshockov) [inline src/Psy/ParserFactory.php]: Fixed.

File: src/Psy/ParserFactory.php (hunk 17 above)
Comment (alexeyshockov) [inline src/Psy/ParserFactory.php]: Fixed.

**Comment by alexeyshockov** [comment]
Fixed all issues, please review.

**Comment by bobthecow** [comment]
Looking good. Thank you!
